generator/recipe_test.go

```diff
@@ -0,0 +1,205 @@
+		return mockProcessor
+	}); err != nil {
+		t.Fatal(err)
+	}
+
+	type args struct {
+		p generator.RecipeParams
+	}
+	tests := []struct {
+		name        string
+		args        args
+		expected    *generator.TemplateData
+		expectedErr string
+	}{
+		{
+			name: "success with minimal params",
+			args: args{
+				p: generator.RecipeParams{
+					Name: "test-name",
+				},
+			},
+			expected: &generator.TemplateData{
+				Name:    "test-name",
+				Version: recipeVersions[len(recipeVersions)-1],
+			},
+		},
+		{
+			name: "success with full params",
+			args: args{
+				p: generator.RecipeParams{
+					Name:       "test-name",
+					Source:     "test-extractor",
+					Sinks:      []string{"test-sink"},
+					Processors: []string{"test-processor"},
+				},
+			},
+			expected: &generator.TemplateData{
+				Name:    "test-name",
+				Version: recipeVersions[len(recipeVersions)-1],
+				Source: struct {
+					Name         string
+					Scope        string
+					SampleConfig string
+				}{
+					Name: "test-extractor",
+				},
+				Sinks: map[string]string{
+					"test-sink": "",
+				},
+				Processors: map[string]string{
+					"test-processor": "",
+				},
+			},
+		},
+		{
+			name: "error with invalid source",
+			args: args{
+				p: generator.RecipeParams{
+					Name:   "test-name",
+					Source: "invalid-source",
+				},
+			},
+			expected:    nil,
+			expectedErr: "provide extractor information: could not find extractor",
+		},
+		{
+			name: "error with invalid sinks",
+			args: args{
+				p: generator.RecipeParams{
+					Name:  "test-name",
+					Sinks: []string{"invalid-sink"},
+				},
+			},
+			expected:    nil,
+			expectedErr: "provide sink information: could not find sink",
+		},
+		{
+			name: "error with invalid processors",
+			args: args{
+				p: generator.RecipeParams{
+					Name:       "test-name",
+					Processors: []string{"invalid-processor"},
+				},
+			},
+			expected:    nil,
+			expectedErr: "provide processor information: could not find processor",
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			actual, err := generator.Recipe(tt.args.p)
+			if tt.expectedErr != "" {
+				assert.ErrorContains(t, err, tt.expectedErr)
+				return
+			}
+			assert.NoError(t, err)
+
+			assert.Equal(t, actual, tt.expected)
+		})
+	}
+}
+
+func TestRecipeWriteTo(t *testing.T) {
+	type args struct {
+		p generator.RecipeParams
+	}
+	tests := []struct {
+		name           string
+		args           args
+		expectedWriter string
+		expectedErr    bool
+	}{
+		{
+			name: "success with minimal params",
+			args: args{
+				p: generator.RecipeParams{
+					Name: "test-name",
+				},
+			},
+			expectedWriter: `name: test-name
+version: v1beta1
+source:
+  name: 
+  config:     
+    
+`,
+			expectedErr: false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			writer := &bytes.Buffer{}
+			if err := generator.RecipeWriteTo(tt.args.p, writer); (err != nil) != tt.expectedErr {
```

Comment:
Can we use assertions here? It will give us consistent error output when something fails.